Error when upgrade/downgrade subscription on Android with DEFERRED mode

### Version of react-native-iap

4.3.3

### Version of react-native

0.59.10

### Platforms you faced the error (IOS or Android or both?)

Android

### Expected behavior

When upgrade/downgrade a subscription with DEFERRED mode, purchase update listener should be called with new transaction receipt.

### Actual behavior

purchase error listener is called with following error:

{ message: 'purchases are null.',
code: 'OK',
debugMessage: '',
responseCode: 0 }

### Tested environment (Emulator? Real Device?)

Real Device - sandbox

### Steps to reproduce the behavior

1.Purchase a subscription with RNIap.requestSubscription(sku), this works correctly.

2.Upgrade or downgrade the subscription with RNIap.requestSubscription(newSku, false, sku, 4). (4 is the DEFERRED Proration Mode) Google Play billing dialog will show the subscription plan is changed successfully, and you will receive a notification email from Google Play about that. However, the error listener is called with the above mentioned error.

3.If you call RNIap.requestSubscription(newSku, false, sku, 4) again, Google Play billing dialog will say they cannot change the subscription plan, and the error listener is called with the following error:

{ message: 'Google is indicating that we have some issue connecting to payment.',
  code: 'E_DEVELOPER_ERROR',
  debugMessage: '',
  responseCode: 5 }

But sometimes Google Play billing dialog will say your order is under processing and your product should be delivered soon, and the error listener is called with the following error:

{ message: 'You already own this item.',
  code: 'E_ALREADY_OWNED',
  debugMessage: '',
  responseCode: 7 }

I guess this is due to the transaction of step 2 is waiting to be acknowledged. However, since step 2 does not return a receipt, we cannot acknowledge it.

@hyochan 4.4.0-rc.2 build failed on:

```
RNIapModule.java:436: error: incompatible types: SkuDetails cannot be converted to String:
   builder.setOldSku(selectedOldSku);
                     ^
```

--- a/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
+++ b/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
@@ -414,7 +414,7 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
         }
         builder.setSkuDetails(selectedSku);
 
-        if (type.equals(oldSku != null) {
+        if (type.equals(oldSku != null)) {
           builder.setOldSku(oldSku);
         }
 

--- a/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
+++ b/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
@@ -414,7 +414,7 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
         }
         builder.setSkuDetails(selectedSku);
 
-        if (type.equals(BillingClient.SkuType.SUBS) && oldSku != null && !oldSku.isEmpty()) {
+        if (type.equals(oldSku != null) {
           builder.setOldSku(oldSku);
         }
 

--- a/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
+++ b/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
@@ -415,7 +415,25 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
         builder.setSkuDetails(selectedSku);
 
         if (type.equals(BillingClient.SkuType.SUBS) && oldSku != null && !oldSku.isEmpty()) {
-          builder.setOldSku(oldSku);
+          SkuDetails selectedOldSku = null;
+          for (SkuDetails skuDetail : skus) {
+            if (skuDetail.getSku().equals(oldSku)) {
+              selectedOldSku = skuDetail;
+              break;
+            }
+          }
+
+          if (selectedOldSku == null) {
+            String debugMessage = "The old sku was not found. Please fetch products first by calling getItems";
+            WritableMap error = Arguments.createMap();
+            error.putString("debugMessage", debugMessage);
+            error.putString("code", PROMISE_BUY_ITEM);
+            error.putString("message", debugMessage);
+            sendEvent(reactContext, "purchase-error", error);
+            promise.reject(PROMISE_BUY_ITEM, debugMessage);
+            return;
+          }
+          builder.setOldSku(selectedOldSku);
         }
 
         if (prorationMode != null && prorationMode != -1) {

--- a/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
+++ b/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
@@ -414,7 +414,7 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
         }
         builder.setSkuDetails(selectedSku);
 
-        if (type.equals(oldSku != null)) {
+        if (oldSku != null) {
           builder.setOldSku(oldSku);
         }
 

--- a/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
+++ b/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
@@ -458,28 +458,34 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
 
   @ReactMethod
   public void acknowledgePurchase(final String token, final String developerPayLoad, final Promise promise) {
-    AcknowledgePurchaseParams acknowledgePurchaseParams =
-            AcknowledgePurchaseParams.newBuilder()
-                    .setPurchaseToken(token)
-                    .setDeveloperPayload(developerPayLoad)
-                    .build();
-    billingClient.acknowledgePurchase(acknowledgePurchaseParams, new AcknowledgePurchaseResponseListener() {
+    ensureConnection(promise, new Runnable(){
       @Override
-      public void onAcknowledgePurchaseResponse(BillingResult billingResult) {
-        if (billingResult.getResponseCode() != BillingClient.BillingResponseCode.OK) {
-          DoobooUtils.getInstance().rejectPromiseWithBillingError(promise, billingResult.getResponseCode());
-        }
-        try {
-          WritableMap map = Arguments.createMap();
-          map.putInt("responseCode", billingResult.getResponseCode());
-          map.putString("debugMessage", billingResult.getDebugMessage());
-          String[] errorData = DoobooUtils.getInstance().getBillingResponseData(billingResult.getResponseCode());
-          map.putString("code", errorData[0]);
-          map.putString("message", errorData[1]);
-          promise.resolve(map);
-        } catch (ObjectAlreadyConsumedException oce) {
-          Log.e(TAG, oce.getMessage());
-        }
+      public void run() {
+        AcknowledgePurchaseParams acknowledgePurchaseParams =
+                AcknowledgePurchaseParams.newBuilder()
+                        .setPurchaseToken(token)
+                        .setDeveloperPayload(developerPayLoad)
+                        .build();
+                        
+        billingClient.acknowledgePurchase(acknowledgePurchaseParams, new AcknowledgePurchaseResponseListener() {
+          @Override
+          public void onAcknowledgePurchaseResponse(BillingResult billingResult) {
+            if (billingResult.getResponseCode() != BillingClient.BillingResponseCode.OK) {
+              DoobooUtils.getInstance().rejectPromiseWithBillingError(promise, billingResult.getResponseCode());
+            }
+            try {
+              WritableMap map = Arguments.createMap();
+              map.putInt("responseCode", billingResult.getResponseCode());
+              map.putString("debugMessage", billingResult.getDebugMessage());
+              String[] errorData = DoobooUtils.getInstance().getBillingResponseData(billingResult.getResponseCode());
+              map.putString("code", errorData[0]);
+              map.putString("message", errorData[1]);
+              promise.resolve(map);
+            } catch (ObjectAlreadyConsumedException oce) {
+              Log.e(TAG, oce.getMessage());
+            }
+          }
+        });
       }
     });
   }

--- a/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
+++ b/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
@@ -415,25 +415,7 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
         builder.setSkuDetails(selectedSku);
 
         if (type.equals(BillingClient.SkuType.SUBS) && oldSku != null && !oldSku.isEmpty()) {
-          SkuDetails selectedOldSku = null;
-          for (SkuDetails skuDetail : skus) {
-            if (skuDetail.getSku().equals(oldSku)) {
-              selectedOldSku = skuDetail;
-              break;
-            }
-          }
-
-          if (selectedOldSku == null) {
-            String debugMessage = "The old sku was not found. Please fetch products first by calling getItems";
-            WritableMap error = Arguments.createMap();
-            error.putString("debugMessage", debugMessage);
-            error.putString("code", PROMISE_BUY_ITEM);
-            error.putString("message", debugMessage);
-            sendEvent(reactContext, "purchase-error", error);
-            promise.reject(PROMISE_BUY_ITEM, debugMessage);
-            return;
-          }
-          builder.setOldSku(selectedOldSku);
+          builder.setOldSku(oldSku);
         }
 
         if (prorationMode != null && prorationMode != -1) {

--- a/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
+++ b/android/src/main/java/com/dooboolab/RNIap/RNIapModule.java
@@ -394,26 +394,6 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
         DoobooUtils.getInstance().addPromiseForKey(PROMISE_BUY_ITEM, promise);
         final BillingFlowParams.Builder builder = BillingFlowParams.newBuilder();
 
-        if (type.equals(BillingClient.SkuType.SUBS) && oldSku != null && !oldSku.isEmpty()) {
-          // Subscription upgrade/downgrade
-          if (prorationMode != null && prorationMode != -1) {
-            builder.setOldSku(oldSku);
-            if (prorationMode == BillingFlowParams.ProrationMode.IMMEDIATE_AND_CHARGE_PRORATED_PRICE) {
-              builder.setReplaceSkusProrationMode(BillingFlowParams.ProrationMode.IMMEDIATE_AND_CHARGE_PRORATED_PRICE);
-            } else if (prorationMode == BillingFlowParams.ProrationMode.IMMEDIATE_WITHOUT_PRORATION) {
-              builder.setReplaceSkusProrationMode(BillingFlowParams.ProrationMode.IMMEDIATE_WITHOUT_PRORATION);
-            } else {
-              builder.setOldSku(oldSku);
-            }
-          } else {
-            builder.setOldSku(oldSku);
-          }
-        }
-
-        if (prorationMode != 0 && prorationMode != -1) {
-          builder.setReplaceSkusProrationMode(prorationMode);
-        }
-
         SkuDetails selectedSku = null;
         for (SkuDetails skuDetail : skus) {
           if (skuDetail.getSku().equals(sku)) {
@@ -421,6 +401,7 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
             break;
           }
         }
+
         if (selectedSku == null) {
           String debugMessage = "The sku was not found. Please fetch products first by calling getItems";
           WritableMap error = Arguments.createMap();
@@ -431,6 +412,35 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
           promise.reject(PROMISE_BUY_ITEM, debugMessage);
           return;
         }
+        builder.setSkuDetails(selectedSku);
+
+        if (type.equals(BillingClient.SkuType.SUBS) && oldSku != null && !oldSku.isEmpty()) {
+          builder.setOldSku(oldSku);
+        }
+
+        if (prorationMode != null && prorationMode != -1) {
+          if (prorationMode == BillingFlowParams.ProrationMode.IMMEDIATE_AND_CHARGE_PRORATED_PRICE) {
+            builder.setReplaceSkusProrationMode(BillingFlowParams.ProrationMode.IMMEDIATE_AND_CHARGE_PRORATED_PRICE);
+            if (type.equals(BillingClient.SkuType.SUBS) == false) {
+              String debugMessage = "IMMEDIATE_AND_CHARGE_PRORATED_PRICE for proration mode only works in subscription purchase.";
+              WritableMap error = Arguments.createMap();
+              error.putString("debugMessage", debugMessage);
+              error.putString("code", PROMISE_BUY_ITEM);
+              error.putString("message", debugMessage);
+              sendEvent(reactContext, "purchase-error", error);
+              promise.reject(PROMISE_BUY_ITEM, debugMessage);
+              return;
+            }
+          } else if (prorationMode == BillingFlowParams.ProrationMode.IMMEDIATE_WITHOUT_PRORATION) {
+            builder.setReplaceSkusProrationMode(BillingFlowParams.ProrationMode.IMMEDIATE_WITHOUT_PRORATION);
+          } else if (prorationMode == BillingFlowParams.ProrationMode.DEFERRED) {
+            builder.setReplaceSkusProrationMode(BillingFlowParams.ProrationMode.DEFERRED);
+          } else if (prorationMode == BillingFlowParams.ProrationMode.IMMEDIATE_WITH_TIME_PRORATION) {
+            builder.setReplaceSkusProrationMode(BillingFlowParams.ProrationMode.IMMEDIATE_WITH_TIME_PRORATION);
+          } else {
+            builder.setReplaceSkusProrationMode(BillingFlowParams.ProrationMode.UNKNOWN_SUBSCRIPTION_UPGRADE_DOWNGRADE_POLICY);
+          }
+        }
 
         if (accountId != null) {
           builder.setAccountId(accountId);
@@ -439,7 +449,6 @@ public class RNIapModule extends ReactContextBaseJavaModule implements Purchases
           builder.setDeveloperId(developerId);
         }
 
-        builder.setSkuDetails(selectedSku);
         BillingFlowParams flowParams = builder.build();
         BillingResult billingResult = billingClient.launchBillingFlow(activity, flowParams);
         String[] errorData = DoobooUtils.getInstance().getBillingResponseData(billingResult.getResponseCode());

--- a/index.ts
+++ b/index.ts
@@ -36,6 +36,14 @@ export enum IAPErrorCode {
   E_BILLING_RESPONSE_JSON_PARSE_ERROR = 'E_BILLING_RESPONSE_JSON_PARSE_ERROR',
 }
 
+export enum ProrationModesAndroid {
+  IMMEDIATE_WITH_TIME_PRORATION = 1,
+  IMMEDIATE_AND_CHARGE_PRORATED_PRICE = 2,
+  IMMEDIATE_WITHOUT_PRORATION = 3,
+  DEFERRED = 4,
+  UNKNOWN_SUBSCRIPTION_UPGRADE_DOWNGRADE_POLICY = 0,
+}
+
 export interface Discount {
   identifier: string;
   type: string;
@@ -225,9 +233,10 @@ export const getSubscriptions = (skus: string[]): Promise<Subscription[]> =>
  * Gets an invetory of purchases made by the user regardless of consumption status
  * @returns {Promise<(InAppPurchase | SubscriptionPurchase)[]>}
  */
-export const getPurchaseHistory = (): Promise<
-  (InAppPurchase | SubscriptionPurchase)[]
-> =>
+export const getPurchaseHistory = (): Promise<(
+  | InAppPurchase
+  | SubscriptionPurchase
+)[]> =>
   Platform.select({
     ios: async () => {
       checkNativeiOSAvailable();
@@ -249,9 +258,10 @@ export const getPurchaseHistory = (): Promise<
  * Get all purchases made by the user (either non-consumable, or haven't been consumed yet)
  * @returns {Promise<(InAppPurchase | SubscriptionPurchase)[]>}
  */
-export const getAvailablePurchases = (): Promise<
-  (InAppPurchase | SubscriptionPurchase)[]
-> =>
+export const getAvailablePurchases = (): Promise<(
+  | InAppPurchase
+  | SubscriptionPurchase
+)[]> =>
   Platform.select({
     ios: async () => {
       checkNativeiOSAvailable();
@@ -318,7 +328,7 @@ export const requestPurchase = (
  * @param {string} sku The product's sku/ID
  * @param {boolean} [andDangerouslyFinishTransactionAutomaticallyIOS] You should set this to false and call finishTransaction manually when you have delivered the purchased goods to the user. It defaults to true to provide backwards compatibility. Will default to false in version 4.0.0.
  * @param {string} [oldSkuAndroid] SKU that the user is upgrading or downgrading from.
- * @param {number} [prorationModeAndroid] UNKNOWN_SUBSCRIPTION_UPGRADE_DOWNGRADE_POLICY, IMMEDIATE_WITH_TIME_PRORATION, IMMEDIATE_AND_CHARGE_PRORATED_PRICE, IMMEDIATE_WITHOUT_PRORATION, DEFERRED
+ * @param {ProrationModesAndroid} [prorationModeAndroid] UNKNOWN_SUBSCRIPTION_UPGRADE_DOWNGRADE_POLICY, IMMEDIATE_WITH_TIME_PRORATION, IMMEDIATE_AND_CHARGE_PRORATED_PRICE, IMMEDIATE_WITHOUT_PRORATION, DEFERRED
  * @param {string} [developerIdAndroid] Specify an optional obfuscated string of developer profile name.
  * @param {string} [userIdAndroid] Specify an optional obfuscated string that is uniquely associated with the user's account in.
  * @returns {Promise<void>}
@@ -327,7 +337,7 @@ export const requestSubscription = (
   sku: string,
   andDangerouslyFinishTransactionAutomaticallyIOS?: boolean,
   oldSkuAndroid?: string,
-  prorationModeAndroid?: number,
+  prorationModeAndroid?: ProrationModesAndroid,
   developerIdAndroid?: string,
   userIdAndroid?: string,
 ): Promise<SubscriptionPurchase> =>

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "react-native-iap",
-  "version": "4.3.4",
+  "version": "4.4.0",
   "description": "React Native In App Purchase Module.",
   "main": "index.js",
   "types": "index.d.ts",

